src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs

```diff
@@ -110,13 +105,20 @@ public static partial class JsonSerializer
                 value = converter.CreateFromList(elementType, (IList)value);
                 setPropertyDirectly = true;
             }
+            else if (!popStackOnEnd)
```

Comment:
Should this check be right up front (before the if/else block)?

Reply:
No the code above with TempEnumerableValues (used when creating arrays) wouldn't work correctly.

This `if` is basically saying if we aren't returning a value (no 'pop' to return a value to the root call stack and return to user) then the "endarray" code doesn't need to do anything since the IList was already populated.